agora-specs/Property/Governor.hs

```diff
@@ -67,172 +93,197 @@ data GovernorDatumCases
-          execute <- validGT
-          create <- validGT
-          toVoting <- validGT
-          vote <- validGT
-          cosign <- validGT
-          le0 <- taggedInteger (-1000, -1)
-
-          case c of
-            ExecuteLE0 ->
-              -- execute < 0
-              return $ ProposalThresholds le0 create toVoting vote cosign
-            CreateLE0 ->
-              -- c < 0
-              return $ ProposalThresholds execute le0 toVoting vote cosign
-            ToVotingLE0 ->
-              return $ ProposalThresholds execute create le0 vote cosign
-            VoteLE0 ->
-              -- vote < 0
-              return $ ProposalThresholds execute create toVoting le0 cosign
-            CosignLE0 ->
-              return $ ProposalThresholds execute create toVoting vote le0
-            Correct ->
-              return $ ProposalThresholds execute create toVoting vote cosign
+        correctCase = \case
+          Correct -> True
+          _ -> False
+
+        classifier :: GovernorDatum -> GovernorDatumCases
+        classifier
+          ( view #proposalThresholds ->
+              ProposalThresholds
+                execute
+                create
+                toVoting
+                vote
+                cosign
+            )
+            | execute < 0 = ExecuteLE0
+            | create < 0 = CreateLE0
+            | toVoting < 0 = ToVotingLE0
+            | vote < 0 = VoteLE0
+            | cosign < 0 = CosignLE0
+            | otherwise = Correct
+
+--------------------------------------------------------------------------------
 
 data GovernorPolicyCases
   = ReferenceUTXONotSpent
   | IncorrectAmountOfTokenMinted
   | GovernorOutputNotFound
-  | GovernorPolicyCorrect
   deriving stock (Eq, Show)
 
-instance Universe GovernorPolicyCases where
-  universe =
-    [ ReferenceUTXONotSpent
-    , IncorrectAmountOfTokenMinted
-    , GovernorOutputNotFound
-    , GovernorPolicyCorrect
-    ]
-
-instance Finite GovernorPolicyCases where
-  universeF = universe
-  cardinality = Tagged 4
-
-governorMintingProperty :: Property
-governorMintingProperty =
-  classifiedPropertyNative gen (const []) expected classifier actual
+governorMintingPolicyTests :: [TestTree]
+governorMintingPolicyTests =
+  [ mkGovMintingCasePropertyTest
+      "Reference input spend test"
+      ReferenceUTXONotSpent
+      "Spent"
+      "Not spent"
+  , mkGovMintingCasePropertyTest
+      "Amount of token minted test"
+      IncorrectAmountOfTokenMinted
+      "Correct"
+      "Incorrect"
+  , mkGovMintingCasePropertyTest
+      "Governor output presense"
+      GovernorOutputNotFound
+      "Present"
+      "Absent"
+  ]
+
+{- | Creates a property by compiling governorPolicy script with given arguments
+   and checking if it runs as expected by a test.
+-}
+governorPolicyValid :: ScriptContext -> Bool -> Property
+governorPolicyValid ctx shouldSuceed =
+  let s = compile deterministicTracingConfig $ governorPolicy' # pconstant ctx
```

Comment:
I _think_ (but I am not certain) this is where we want pre-compilation. 

I.e., you want `governorPolicy'` already compiled, you script context already compiled, etc -- and then you want to use `applyArguments` from `PlutusLegerApi.V1.Scripts`.

I think this will greatly increase the speed.

@kozross, can you chime in?

Reply:
This should be addressed already by the use of `loudEval`.

Reply:
done